barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_trace.cpp

```diff
@@ -66,50 +66,94 @@ void AvmTraceBuilder::reset()
 
     external_call_counter = 0;
 }
-
-AvmTraceBuilder::IndirectThreeResolution AvmTraceBuilder::resolve_ind_three(
-    uint8_t space_id, uint32_t clk, uint8_t indirect, uint32_t a_offset, uint32_t b_offset, uint32_t c_offset)
+/**
+ * @brief Returns an array of mem_offsets and tags them with their given Addressing Mode (direct/indirect) based on the
+ * given indirect byte.
+ * @tparam N The number of memory offsets to resolve.
+ */
+template <size_t N> std::array<ResolvedAddr, N> resolve_indirects(uint8_t indirect, std::array<uint32_t, N> mem_offsets)
```

Comment:
wait, IIUC the name is misleading. I thought this would be doing the same as the `Addresing.fromWire().resolve()` but it does not resolve the addresses, it only returns the same address, with the indirection bit. It's more of an "unpacking" of the indirects.

I see that given how the circuit works, you might need to do the resolution later in, e.g., indirect_read_and_load_from_memory. A pity if that's so. But then I'd change the name because it clashes with the semantics we give it in TS.

Reply:
Yep that's correct. The issue right now is that if we resolve an indirect address, we create a row in the mem trace which will require aligning of `clks` and is fraught with ordering issues.

We may be able to resolve this in the future by also exposing an `unconstrained_read..` version which allows us access to the memory laying down constraints.